### PR TITLE
Feature/auth refactor

### PR DESCRIPTION
**Pull request by daiLlew**
### What

Refactored CMD dataset permissions endpoint. 
- Split functionality into 2 endpoints `ServiceDatasetPermissions` and `UserDatasetPermissions`.
- Valid Admin/Editor users are permitted full `CRUD` permissions for all datasets. For these users the  collection ID param is ignored. However, they must provide a dataset ID - so Zebedee can log what they were given access to.
- Viewer users are granted `R` permissions if they have `view` access to a collection **and** the collection contains the dataset.

### CODE REVIEW

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/cmd/PermissionsAPIBase.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+        try {
+            CRUD permissions = getPermissions(request, response);
+            writeResponse(response, permissions, SC_OK);
+        } catch (PermissionsException ex) {
+            error().exception(ex).log("error getting dataset permissions");
+            writeResponse(response, new Error(ex.getMessage()), ex.statusCode);
+        } catch (IOException ex) {
+            error().exception(ex).log("internal sever error failed to write response body");
+            response.setStatus(SC_INTERNAL_SERVER_ERROR);
```
Comment (mikeAdamss) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/cmd/PermissionsAPIBase.java]: Do you need to write this?

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/authorisation/AuthorisationServiceImpl.java [hunk 2]
```diff
@@ -14,15 +11,10 @@
 import org.apache.commons.lang3.StringUtils;
 
 import java.io.IOException;
+import java.util.function.Function;
```
Comment (mikeAdamss) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/authorisation/AuthorisationServiceImpl.java]: Is this used?

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/configuration/CMSFeatureFlags.java [hunk 3]
```diff
@@ -19,11 +20,14 @@
...
+    private final boolean isCMDAuthEnabled;
+
     /**
      * Construct a new feature flags instance.
      */
     private CMSFeatureFlags() {
         this.isDatasetImportEnabled = Boolean.valueOf(getConfigValue(ENABLE_DATASET_IMPORT));
+        this.isCMDAuthEnabled = Boolean.valueOf(getConfigValue(ENABLE_CMD_AUTH));
 
         info().data(ENABLE_DATASET_IMPORT, isDatasetImportEnabled).log("CMS feature flags configurations");
```
Comment (mikeAdamss) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/configuration/CMSFeatureFlags.java]: is it worth logging if the new flag is on or off?

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/permissions/cmd/CMDPermissionsServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+import static com.github.onsdigital.zebedee.permissions.cmd.PermissionsException.serviceTokenNotProvidedException;
+import static com.github.onsdigital.zebedee.permissions.cmd.PermissionsException.sessionExpiredException;
+import static com.github.onsdigital.zebedee.permissions.cmd.PermissionsException.sessionIDNotProvidedException;
+import static com.github.onsdigital.zebedee.permissions.cmd.PermissionsException.sessionNotFoundException;
+import static org.apache.commons.lang3.StringUtils.isEmpty;
+import static org.apache.http.HttpStatus.SC_BAD_REQUEST;
+
+public class CMDPermissionsServiceImpl implements CMDPermissionsService {
+
+    private static final String DATASET_ID_NOT_PROVIDED = "no permissions granted to user dataset ID required but not" +
```
Comment (mikeAdamss) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/permissions/cmd/CMDPermissionsServiceImpl.java]: is this one used?

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/cmd/PermissionsAPIBase.java (hunk 1 above)
Comment (daiLlew) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/api/cmd/PermissionsAPIBase.java]: Yes. The default would be a 200.

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/authorisation/AuthorisationServiceImpl.java (hunk 2 above)
Comment (daiLlew) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/authorisation/AuthorisationServiceImpl.java]: not anymore will remove.

**Pull request merged**